File: SexyBiscuit.Engine/Gameplay/GameState.cs
using SexyBiscuit.Engine.Core;

namespace SexyBiscuit.Engine.Gameplay;

/// <summary>
/// Match-wide state that every client needs — elapsed time, the player list, the current
/// phase. Modelled on Unreal's <c>AGameStateBase</c>.
/// </summary>
/// <remarks>
/// The split from <see cref="GameMode"/> is deliberate: the game mode holds the rules and
/// exists only on the server, while the game state holds the facts and is visible everywhere.
/// Anything a client's HUD needs to read belongs here, not on the mode.
/// </remarks>
public class GameState : Actor
{
    /// <summary>Seconds since the match entered <see cref="MatchState.InProgress"/>.</summary>
    public float ElapsedTime { get; internal set; }

    /// <summary>Current phase of the match.</summary>
    public MatchState MatchState { get; internal set; } = MatchState.WaitingToStart;

    /// <summary>Every player currently in the session, including bots.</summary>
    public List<PlayerState> Players { get; } = new();

    /// <summary>Raised when <see cref="MatchState"/> changes, with the new state.</summary>
    public SBEvent<MatchState> MatchStateChanged { get; } = new();

    public GameState() : base("GameState") { }

    /// <summary>Adds a player to <see cref="Players"/> if not already present.</summary>
    public void AddPlayer(PlayerState player)
    {
        if (!Players.Contains(player)) Players.Add(player);
    }

    /// <summary>Removes a player from <see cref="Players"/>.</summary>
    public void RemovePlayer(PlayerState player) => Players.Remove(player);

    /// <summary>Players sorted by score, highest first.</summary>
    public IEnumerable<PlayerState> GetScoreboard()
        => Players.OrderByDescending(p => p.Score);

    protected override void Update(float dt)
    {
        if (MatchState == MatchState.InProgress) ElapsedTime += dt;
    }
}

/// <summary>Phases a match moves through, in order.</summary>
public enum MatchState
{
    /// <summary>Players are still connecting; the mode has not started the match.</summary>
    WaitingToStart,

    /// <summary>Gameplay is live.</summary>
    InProgress,

    /// <summary>Gameplay is suspended but the match has not ended.</summary>
    Paused,

    /// <summary>The match finished; a winner may be known.</summary>
    Ended,
}

File: html5/src/gameplay/GameState.js
// -----------------------------------------------------------------------------
// GameState — match-wide state every client can see.
// -----------------------------------------------------------------------------

import { Actor } from '../core/Actor.js';
import { registerActor } from '../core/TypeRegistry.js';
import { SBEvent } from '../core/SBEvent.js';

/** Where a match is in its lifecycle. */
export const MatchState = Object.freeze({
    WaitingToStart: 'WaitingToStart',
    InProgress: 'InProgress',
    Paused: 'Paused',          // the C# enum has it; a state file crosses engines with it
    Ended: 'Ended',
});

/** The shared state of the match in progress. */
export class GameState extends Actor {
    constructor(name = 'GameState') {
        super(name);
        this.elapsedTime = 0;
        this._matchState = MatchState.WaitingToStart;

        /** @type {import('./PlayerState.js').PlayerState[]} */
        this.players = [];

        this.matchStateChanged = new SBEvent();
    }

    get matchState() { return this._matchState; }
    set matchState(value) {
        if (this._matchState === value) return;
        this._matchState = value;
        this.matchStateChanged.broadcast(value);
    }

    addPlayer(playerState) {
        if (!playerState || this.players.includes(playerState)) return;
        this.players.push(playerState);
    }

    removePlayer(playerState) {
        const i = this.players.indexOf(playerState);
        if (i >= 0) this.players.splice(i, 1);
    }

    /** Players sorted by score, highest first. */
    getScoreboard() { return [...this.players].sort((a, b) => b.score - a.score); }

    update(dt) {
        if (this._matchState === MatchState.InProgress) this.elapsedTime += dt;
    }
}
registerActor(GameState, { category: 'Gameplay', summary: 'Match-wide shared state.' });
